### DISCUSSION TITLE
Rolling multiple database operation in a Transaction

### DISCUSSION THREAD

**Discussion by henda79** [Azure/azure-mobile-apps#406] [Q&A]
I'm looking for some advice on best practices for rolling multiple DB operations in to a transaction. I would have thought others would already use something like this as it's a common scenario right ?

So if I want to update across several tables, or insert or whatever and something fails for whatever reason, I would like to rollback the changes.

In my case this would be on the client in an offline situation.

What is the best way to go about this ?

**Comment by adrianhall** [comment]
This is not the right tool for that job.  We don't support transactions at all.

**Reply by henda79** [reply]
Surely it would be beneficial to support this to ensure the integrity of the data ?

**Reply by adrianhall** [reply]
One of the underlying premises of Azure Mobile Apps is that the tables are distinct and (importantly) can be independently synchronized.  If you need transactions, you are breaking that underlying premise.

I'm not saying it's a bad idea - just that adding support for transactions and roll-backs would be a huge effort that explicitly changes the underpinnings of the library and would cause a complete re-write of the offline store model.  At that point, we'd be re-creating Entity Framework, which is definitely not something we want to do.

**Reply by henda79** [reply]
Yes, I see what you're saying. I should have really thought about this more before refactoring. I'm fairly sure I'm going to need transactions to protect the data as I know some of the functions performed on the client require updating several tables and the data includes financial transactions

**Reply by adrianhall** [reply]
I've seen multiple devs deal with this in a live manner, especially when considering financial transactions - there is an API on the backend that does any comple update - the API updates all the tables at once in a transaction and returns a 200 for success, or 409 for failure (Conflict - return a JSON document describing the problem - code/error/etc.)  Now you can pull the tables to get the result of the transaction.

You can still queue these transactions, but you can't update the offline database until you are online (generally a requirement for financial transactions).

**Reply by henda79** [reply]
Yeah that's not going to work in our case. In the field we don't always have a connection (as you well know) but will still need to perform the transaction and give the customer a reciept (its not the usual like a customer purchasing something). So we can't rely on a online API unfortunately.

And since we can't control the order the data is pushed to the server, I suppose we can't have referential data on the server side either :(

**Reply by adrianhall** [reply]
The order that records are pushed to the server is the same as injected into the queue - with three exceptions:

1) You use the new multi-threading code for pushes
2) You push single tables (in which case the ordering in the table is preserved, but not the ordering between tables)
3) Collapsed records

That item 3 may need some explaining.  Let's say you do this:

* Create record 1 
* Create record 2 
* Update record 1 

There are only two queue entries - in this order:

* Create record 1 (modified)
* Create record 2 

That's because the update has been collapsed into the create operation of the same record.  You can do the same thing with updates (two updates becomes one update) and deletes (create-delete is removed from the queue, and update-delete becomes a delete operation at the point of the delete and the update is removed)

**Reply by henda79** [reply]
Thanks for the detailed explanation Adrian. Glad you elaborated on #3, as I would have needed clarification of that.

When you say use the new multi-threading code, is this something in the pipeline or have I missed something ?

**Reply by adrianhall** [reply]
It's checked into main, but not released yet.  There is a new ParallelOperations option that allows you to parallelize requests made as part of the push queue - you can use up to 8 threads (but may be limited by the platform - Android, for example, only allows 4 outbound HTTP requests at a time).

**Reply by henda79** [reply]
I'll probably hang until release and maybe a sample, I've still got quite a bit of refactoring to do. If I get done before the official release I'll pull the changes and have a look.

**Reply by henda79** [reply]
@adrianhall , so I've come full circle to transactions again as I need to be able to rollback any issues. Maybe I was not clear before, and that I am attempting this on the offline database.

With that said, I notice in `OfflineSQLiteStore.cs` under the `UpsertAsync()` method there is a block of code which performs inserts / updates within a transaction.

```
            using (operationLock.AcquireLock())
            {
                ExecuteNonQueryInternal("BEGIN TRANSACTION");
                BatchInsert(tableName, items, columns.Where(c => c.IsIdColumn).Take(1).ToList());
                BatchUpdate(tableName, items, columns);
                ExecuteNonQueryInternal("COMMIT TRANSACTION");
            }
```

This is basically what I was looking for. So I created a subclass so I could use `ExecuteNonQueryInternal`.

However I got an issue as it not possible to have a nested transaction.

Any ideas of a way around this ?

**Reply by adrianhall** [reply]
Azure Mobile Apps does not support transactions.  The offline store you are referencing has a transaction as an implementation detail - it isn't exposed as a transaction to the user (and other implementations of the offline store may not use it).

If you need transactions, Azure Mobile Apps is not the solution for you.  You need to look (generally) at custom solutions.

**Reply by henda79** [reply]
Thanks for the reply. We were using a custom solution but it was costly for our use so decided to use something else and after looking for sometime I could not find anything else other than this.

Now since sqlite supports transactions, I can't see why we can't use them. If I can come up with a simple way which does not break current implementations can I submit a PR ?

**Reply by adrianhall** [reply]
Your problem is that the transactions don't survive the round trip to the server - the server implementation (and, more importantly, the protocol between the client and server) have an implicit reliance on a lack of transactions.  Thus, you can put them in the client but then that transaction is broken across the client-server boundary.

**Reply by henda79** [reply]
I only want to support the transactions on the client, at least that way there is some integrity before it all gets synced with the server.

**Reply by adrianhall** [reply]
Just fork the offline store then.   The correct internals should be available to you to add a transaction as an additional set of methods.

**Reply by henda79** [reply]
Yes, this is what I've done, but If I start a transaction in my code with BEGIN, when I insert the data to the offline database, your code also tries to start another transaction which causes an error, so what I've done for now is just comment out the BEGIN and COMMIT in your code.

**Reply by henda79** [reply]
Where would it be best to have a method to batch the inserts / updates for the offline store.

**Reply by adrianhall** [reply]
first, create a `class OfflineStoreWithMyMods : OfflineSQLiteStore`, then add a method directly to there.  You should be able to use any of the ExecuteQueryInternal methods then.

I think a case can be made for adding something like `ExecuteBatchedSqlInternal()`.  Something like:

``` csharp
struct SqlStatementPackage {
  public string sqlStatement { get; set; }
  public IDictionary<string, object> parameters { get; set; }
}
```

then add:

``` csharp
protected IEnumerable<SqlResult> ExecuteBatchedSqlInternal(IEnumerable<SqlStatementPackage> sqlStatements);
```

This would start a transaction, execute each statement in order, then complete the transaction, returning a SqlResult (which has not been defined, but returns what each of the SQL statement results are plus the result of the transaction).  I'd take a PR for that (with tests, obviously)

**Reply by henda79** [reply]
I've just noticed that `GetTablesAsync` does not return any tables. I am not manually defining them since you introduced the deferred method. In this case we would need a `GetRemoteTablesAsync`

Is this possible or should I define each table ?

**Reply by adrianhall** [reply]
There is no API method to return a list of the remote tables, and GetTablesAsync only returns tables that have been previously defined (i.e. exist in the SQL database).  

Eventually, you have to define the tables you are using.  There is just no way around it.

**Reply by henda79** [reply]
> There is no API method to return a list of the remote tables, and GetTablesAsync only returns tables that have been previously defined (i.e. exist in the SQL database).
> 
> Eventually, you have to define the tables you are using. There is just no way around it.

The tables are present in the offline Database, I just don't explicitly define them as they get defined when they are used, but if the tables are not "used" then the method does not return anything.

Anyway not a major problem for now, I might look at getting tables from the server at a later date which allows us to sync from a fresh install.

Regarding comment above about the batch inserts / updates, I've started on this but there is no way to use it from a higher level with in the app without changes to OfflineTable and SyncContext that is if we still want to use something like

```
await _client!
    .GetOfflineTable<T>()
    .InsertItemsAsync(items)
    .ConfigureAwait(false);
```

**Reply by adrianhall** [reply]
Correct - that's because not all stores have batching and transactions.  There will never be transactions at the higher level.

**Reply by henda79** [reply]
I've been thinking about this and its a lot of work to get it to work, however a quick fix would be to update the code in the UpsertAsync method.

```
            using (operationLock.AcquireLock())
            {
                string savePoint = Guid.NewGuid().ToString("N");
                ExecuteNonQueryInternal($"SAVEPOINT '{savePoint}'");
                BatchInsert(tableName, items, columns.Where(c => c.IsIdColumn).Take(1).ToList());
                BatchUpdate(tableName, items, columns);
                ExecuteNonQueryInternal($"RELEASE '{savePoint}'");
            }
```

Using this would allow me to create a transaction higher up using ExecuteNonQueryInternal without hitting nesting issues.

What do you think ?

**Reply by adrianhall** [reply]
Savepoints have to be used within the context of a transaction.  There is also no point to a savepoint if your batch doesn't rollback at some point.  We'd have to thread the savepoint through the entire system (batch insert and batch update as well).  I'm having a hard time understanding how this would be a useful change higher up in the code.

**Reply by henda79** [reply]
I was under the impression a savepoint created a transaction if none was explicitly created.

In my code, data is collected and is entered across several tables, so there is a risk of having orphaned data if something goes wrong. So what I do is:

```
List<Func<Task>> actions = new();

actions.Add(() => DataService.Insert(myDataForTable1));
actions.Add(() => DataService.Insert(myDataForTable2));
...

await DataService.RunInTransaction(actions).ConfigureAwait(false);
...
        public class OfflineSqlStore : OfflineSQLiteStore
        {
            public OfflineSqlStore(string connectionString) : base(connectionString)
            {
            }

            public async Task RunInTransaction(List<Func<Task>> actions)
            {
                ExecuteNonQueryInternal("BEGIN");

                try
                {
                    foreach (Func<Task> item in actions)
                    {
                        await item.Invoke().ConfigureAwait(false);
                    }

                    ExecuteNonQueryInternal("COMMIT");
                }
                catch
                {
                    ExecuteNonQueryInternal("ROLLBACK");
                    throw;
                }
            }
    }
```

This code did not work as you were also creating an explicit transaction in UpsertAsync however it seems to work when using SAVEPOINT.

I don't wrap everything in a transaction as some changes only hit one table, but in cases where I am inserting a lot of data over several tables then it would be mad not to use a transaction.

**Comment by amertlich** [comment]
@henda79 did you end up finding a workaround for this problem?


> There is also no point to a savepoint if your batch doesn't rollback at some point

@adrianhall I think ☝️is true even with begin/commit transactions, which was addressed here: https://github.com/Azure/azure-mobile-apps/issues/749

I'm not seeing a reason why you couldn't substitute the `BEGIN TRANSACTION` [here](https://github.com/Azure/azure-mobile-apps/blob/main/sdk/dotnet/src/Microsoft.Datasync.Client.SQLiteStore/OfflineSQLiteStore.cs#L285) instead with a `SAVEPOINT data_sync_upsert` and switch `COMMIT` to `RELEASE data_sync_upsert` and then in the exceptional case you `ROLLBACK TO SAVEPOINT data_sync_upsert` rather than just `ROLLBACK` -- i.e like what @henda79 mentions above?

**Reply by henda79** [reply]
I maintain a local copy of the repository and made changes to the code as I outlined above, I don't see any sense in the way it was implemented originally but implementing as above allows me to rollback when things inevitably go wrong.
